### PR TITLE
B/add continuous background job handler

### PR DESCRIPTION
**Pull request by Alexander-Dubrawski**
This PR is the secound of four which aims to refactor the handling of Hyrise jobs (it aims to split #664). Hyrise jobs are interacting directly with the Hyrise to get the meta information and managing the Hyrise settings.  We have three kinds of jobs: 

1. jobs that need to be continuously in the background (for example: get the system information)
2. jobs that will be done asynchronously (for example: loading the tables)
3. jobs that are done synchronously (for example: sending a SQL query over the SQL endpoint)

Both continuous and asynchronous jobs are executed via the apscheduler (https://apscheduler.readthedocs.io/en/stable/). Esspacillay with the asynchronous jobs that's not necessary and increases the complexity (we could just use normal python threads). All three kinds of jobs are scattered in the database and background_scheduler component. The goal of the refactoring is now to have a clear separation for the jobs in the three categories (continuous, asynchronous, and synchronous) and define handlers for that. Moreover, we can stop the continuous jobs while loading/deleting tables (hyrise/hyrise#2180). 

This PR implements the handler for continuous jobs.

## ToDo

- [x] improve docstrings
- [x] move background scheduler logic to database

### CODE REVIEW

File: hyrisecockpit/database_manager/continuous_job_handler.py [hunk 1]
```diff
@@ -22,47 +21,59 @@
...
-        workload_drivers,
+        database_blocked: Value,
     ):
-        """Initialize BackgroundJobManager object."""
-        self._database_id: str = database_id
+        """Initialize continuous Job Handler.
+
+        Args:
+            connection_factory: An object to create a connection to the Hyrise
+                database. All connection relevant information (port, host) is
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/continuous_job_handler.py]: All information relevant for the connection...

File: hyrisecockpit/database_manager/continuous_job_handler.py [hunk 2]
```diff
@@ -22,47 +21,59 @@
...
+                saved in this object.
+            hyrise_active: Flag stored in a shared memory map. This flag
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/continuous_job_handler.py:46]: responsive is different from active. Please rename or clarify.

File: hyrisecockpit/database_manager/continuous_job_handler.py [hunk 3]
```diff
@@ -22,47 +21,59 @@
...
+                stores if the Hyrise instance is responsive or not.
+            worker_pool: An object that manages the queue and task/queue workers. Its
+                api is used to get the queue length.
+            storage_connection_factory: An object to create a connection to the Influx
+                database. All connection relevant information (port, host) is
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/continuous_job_handler.py]: see above

File: hyrisecockpit/database_manager/continuous_job_handler.py [hunk 4]
```diff
@@ -153,7 +164,11 @@ def start(self) -> None:
         self._scheduler.start()
 
     def close(self) -> None:
-        """Close background scheduler."""
+        """Close background scheduler.
+
+        Here we remove all jobs from the background scheduler and
+        shutting it down.
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/continuous_job_handler.py]: `shut`

File: hyrisecockpit/database_manager/database.py [hunk 5]
```diff
@@ -19,7 +20,14 @@
...
 
 class Database(object):
-    """Represents database."""
+    """Represents database.
+
+    The database class is a representation of a registered Hyrise instance.
+    All the communication with the Hyrise is managed by this class and its API.
+    A worker pool object that manages the queue and task/queue workers to put the
+    Hyrise under pressure is part of this class. This class also initializes the
+    influx database for the Hyrise instance.
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: an influx database for this Hyrise instance.

File: hyrisecockpit/database_manager/database.py [hunk 6]
```diff
@@ -31,16 +39,40 @@ def __init__(
...
         storage_host: str,
         storage_password: str,
         storage_port: str,
         storage_user: str,
     ) -> None:
-        """Initialize database object."""
+        """Initialize database object.
+
+        Args:
+            id: A name representation of the Hyrise instance.
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: Sounds odd.

File: hyrisecockpit/database_manager/database.py [hunk 7]
```diff
@@ -31,16 +39,40 @@ def __init__(
...
+            user: The user that connects to the Hyrise instance.
+            password: The password needed to connects to the Hyrise instance.
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: `to connect`

File: hyrisecockpit/database_manager/database.py [hunk 8]
```diff
@@ -31,16 +39,40 @@ def __init__(
...
+            host: The address of the host machine where the Hyrise instance is running.
+            port: The port on which the Hyrise instance is running.
+            dbname: The database name of the Hyrise instance.
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: How does this differ from `id`. This should be clarified.

File: hyrisecockpit/database_manager/database.py [hunk 9]
```diff
@@ -31,16 +39,40 @@ def __init__(
...
+            number_workers: Number of task workers to put the Hyrise instance under
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py:56]: Under pressure sounds too specific. Workers just send queries to a hyrise instance, don't they?

File: hyrisecockpit/database_manager/database.py [hunk 10]
```diff
@@ -31,16 +39,40 @@ def __init__(
...
+                pressure. Every worker is running in its own process.
+            workload_publisher_url: URL which is publishing the workload to the workers.
+                This URL is used by the queue worker to connect to the workload
+                generator via zeromq.
+            storage_host: Address in which the influx database is running.
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: Address of the influx database instance

File: hyrisecockpit/database_manager/database.py [hunk 11]
```diff
@@ -75,20 +107,31 @@ def __init__(
...
-        self._background_scheduler.start()
+        self._continuous_job_handler.start()
 
     def _initialize_influx(self) -> None:
-        """Initialize Influx database."""
+        """Initialize Influx database.
+
+        We drop the database inside influx that has the same name like the
+        database id (Hyrise). We do that to remove all the data from previous
+        runs. Than we create a new database inside influx with the database id
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: Then

File: hyrisecockpit/database_manager/database.py [hunk 12]
```diff
@@ -75,20 +107,31 @@ def __init__(
...
+        (Hyrise). After that we create continuous query that the influx is running
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: a continuous

File: hyrisecockpit/database_manager/database.py [hunk 13]
```diff
@@ -75,20 +107,31 @@ def __init__(
...
+        every x seconds. For example it will so automatically calculate the throughput
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: For example, to automatically calculate...

File: hyrisecockpit/database_manager/database.py [hunk 14]
```diff
@@ -168,11 +211,23 @@ def _initialize_influx(self) -> None:
...
+        We return the number of task that still need to be send to the
+        Hyrise instance.
+        """
         return self._worker_pool.get_queue_length()
 
     def load_data(self, workload: Dict) -> bool:
-        """Load pre-generated tables."""
+        """Load pre-generated tables.
+
+        First we check if the workload and scale factor is valid.
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: First,

File: hyrisecockpit/database_manager/database.py [hunk 15]
```diff
@@ -168,11 +211,23 @@ def _initialize_influx(self) -> None:
...
+        For that workload need to have an equivalent driver and the
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: Not sure: For that workload, we need...?

File: hyrisecockpit/database_manager/database.py [hunk 16]
```diff
@@ -168,11 +211,23 @@ def _initialize_influx(self) -> None:
...
+        driver needs to support the scale factor. Moreover the worker pool
+        needs to be closed. If one of this requirements is not met the function
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: ... is not met,...

File: hyrisecockpit/database_manager/database.py [hunk 17]
```diff
@@ -184,12 +239,20 @@ def load_data(self, workload: Dict) -> bool:
...
-        """Delete tables."""
+        """Delete tables.
+
+        First we check if the workload and scale factor is valid.
+        For that workload need to have an equivalent driver and the
+        driver needs to support the scale factor. Moreover the worker pool
+        needs to be closed. If one of this requirements is not met the function
+        will return false. Otherwise the tables will be deleted via the
+        asynchronous job handler.
+        """
```
Comment (Bensk1) [inline hyrisecockpit/database_manager/database.py]: Is this comment an exact duplicate of the one of `load_data`?

File: tests/database_manager/test_continous_job_handler.py [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+from hyrisecockpit.database_manager.job.update_workload_operator_information import (
+    update_workload_operator_information,
+)
+from hyrisecockpit.database_manager.job.update_workload_statement_information import (
+    update_workload_statement_information,
+)
+
+
+class TestContinuousJobHandler:
+    """This class test the continuous job handler."""
```
Comment (Bensk1) [inline tests/database_manager/test_continous_job_handler.py]: tests

File: tests/database_manager/test_continous_job_handler.py [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+            == "storage_connection_factory"
+        )
+        assert continuous_job_handler._previous_system_data == {
+            "previous_system_usage": None,
+            "previous_process_usage": None,
+        }
+        assert continuous_job_handler._previous_chunk_data == {
+            "value": None,
+        }
+        assert continuous_job_handler._scheduler == mock_background_scheduler_obj
```
Comment (Bensk1) [inline tests/database_manager/test_continous_job_handler.py:56]: Are these asserts the correct way for testing? If the asserts fail, the program crashes, doesn't it? I expected something like `assertEqual`

**Review by Bensk1 — CHANGES_REQUESTED** [review]
Mostly comments regarding small typos in the documentation.

File: tests/database_manager/test_continous_job_handler.py (hunk 19 above)
Comment (Alexander-Dubrawski) [reply tests/database_manager/test_continous_job_handler.py:56]: This is the normal syntax for `pytest`. `pytest` is a third-party unit test framework with a more lighter-weight syntax. If an assert fails only the test fails.

**Review by Bensk1 — APPROVED** [review]

**Pull request merged**